docs/dev_docs.md

```diff
@@ -1,3 +1,449 @@
+
+All of the above can be applied to your environment with:
+
+```
+from pettingzoo.utils import OrderEnforcingWrapper
+from pettingzoo.butterfly import pistonball_v3
+env = pistonball_v3.env()
+env = OrderEnforcingWrapper(env)
+```
+
+## Developer Utils
+
+The utils directory contains a few functions which are helpful for debugging environments. These are documented in the API docs.
+
+The utils directory also contain some classes which are only helpful for developing new environments. These are documented below.
+
+### Agent selector
+
+The `agent_selector` class steps in a cycle
+
+It can be used as follows to cycle through the list of agents:
+
+```python
+from pettingzoo.utils import agent_selector
+agents = ["agent_1", "agent_2", "agent_3"]
+selector = agent_selector(agents)
+agent_selection = selector.reset()
+# agent_selection will be "agent_1"
+for i in range(100):
+    agent_selection = selector.next()
+    # will select "agent_2, agent_3, agent_1, agent_2, agent_3, ..."
+```
+
+### Depreciated Module
+
+The DepreciatedModule is used in PettingZoo to help guide the user away from old obsolete environment versions and to new ones. If you wish to create a similar versioning system, this may be helfpul.
+
+For example, when the user tries to import the `prospector_v0` environment, they import the following variable (defined in `pettingzoo/butterfly/__init__.py`):
+```
+from pettingzoo.utils.depreciated_module import DepreciatedModule
+prospector_v0 = DepreciatedModule("prospector", "v0", "v3")
+```
+This declaration tells the user that `prospector_v0` is depreciated and `prospector_v3` should be used instead. In particular, it gives the following error:
+```
+from pettingzoo.butterfly import prospector_v0
+prospector_v0.env()
+# pettingzoo.utils.depreciated_module.DeprecatedEnv: prospector_v0 is now depreciated, use prospector_v3 instead
+```
+
+## Tests
+
+PettingZoo has a number of tests which it puts environments through. If you are adding a new environment, we encourage you to run these tests on your own environment.
+
+### API test
+
+PettingZoo's API has a number of features and requirements. To make sure your environment is consistent with the API, we have the api_test. Below is an example:
+
+```
+from pettingzoo.test import api_test
+from pettingzoo.butterfly import pistonball_v3
+env = pistonball_v3.env()
+api_test(env, num_cycles=10, verbose_progress=False)
+```
+
+As you can tell, you simply pass an environment to the test. The test will assert or give some other error on an API issue, and will return normally if it passes.
+
+The optional arguments are:
+
+*  `num_cycles`: runs the environment for that many cycles and checks that the output is consistent with the API.
+* `verbose_progress`: Prints out messages to indicate partial completion of the test. Useful for debugging environments.
+
+### Parallel Play test
```

Comment:
Shouldn't it just be called Parallel API Test?

Reply:
done